File: statQuest/PCA.r
# Generate a fake dataset that we can use in the demonstration
# A matrix of data with 10 samples where we measured 100 genes in each sample
data.matrix <- matrix(nrow=100, ncol=10)

# name the samples
# The first 5 samples will be "wt" or "wild type" samples
# "wt" samples are the normal, every day samples
# The last 5 samples will be "ko" or "knock-out" samples
# These are samples that are missing a gene because we knocked it out
colnames(data.matrix) <- c(
  paste("wt", 1:5, sep=""),
  paste("ko", 1:5, sep="")
)

# name the genes
# Usually you'd have things like "Sox9" AND "Utx"
# but since this is a fake dataset, we have gene1, gene2, ..., gene100
rownames(data.matrix) <- paste("gene", 1:100, sep="")

# This is where we give the fake genes fake read counts
# rpois : used poisson distribution
for (i in 1:100) {
  wt.values <- rpois(5, lambda=sample(x=10:1000, size=1))
  ko.values <- rpois(5, lambda=sample(x=10:1000, size=1))
  
  data.matrix[i, ] <- c(wt.values, ko.values)
}

# First 6 rows in our data matrix
# NOTE : The samples are columns, and the genes are rows
head(data.matrix)

# Call prcomp() to do PCA on our data
# The goal is to draw a graph that shows how the samples are related (or not related) to each other
# NOTE : By default, prcomp() expects the samples to be rows and the genes to be columns
# Since the samples in our data matrix are columns, and the genes(variables) are rows
# ★ we have to transpose the matrix using the t() function
# If we don't transpose the matrix,
# we will ultimately get a graph that shows how the genes are related to each other
pca <- prcomp(t(data.matrix), scale=TRUE)

# ★ prcomp() returns three things: x, sdev, rotation

# x contains the principal components (PCs) for drawing a graph
# Here we are using the first two columns in x to draw a 2-D gaph
# that uses the first two PCs
# Remember! Since there are 10 samples, there are 10 PCs
# The first PC accounts for the most variation in the original data
# (the gene expression across all 10 samples),
# the 2nd PC accounts for the second most variation and so on
# To plot a 2-D PCA graph, we usually use the first 2 PCs
# However, sometimes we use PC2 and PC3
plot(pca$x[,1], pca$x[,2])

# Graph interpretation
# 5 of the samples are on one side of the graph
# and the other 5 samples are on the other side of the graph
# To get a sense of how meaningful these clusters are,
# let's see how much variation in the original data PC1 accounts for

# To do this, we use the square of "sdev"
# which stands for "standard deviation",
# to calculate how much variation in the original data each principal component accounts for
pca.var <- pca$sdev^2

# Since the percentage of variation that each PC accounts for is way more interesting
# than the actual value, we calculate the percentages
pca.var.per <- round(pca.var/sum(pca.var)*100, 1)

# Plot the percentages is easy with barplot()
barplot(pca.var.per, main="Scree Plot", xlab="Principal Component", ylab="Percent Variation")

# Graph interpretation
# PC1 accounts for almost all of the variation in the data!
# This means that there is a big difference between two clusters
# (5 on the left and 5 on the right)

# We can use ggplot2 to make a fancy PCA plot that looks nice
# and also provides us with tons of information
install.packages("ggplot2")
library(ggplot2)

# First, format the data the way ggplot2 likes it
# We made a data frame with one column with the sample ids,
# two columns for the X and Y coordinates for each sample
pca.data <- data.frame(Sample=rownames(pca$x),
                       X=pca$x[,1],
                       Y=pca$x[,2])

# Here's what the dataframe looks like
# We have one row per sample
# Each row has a sample ID and X/Y coordinates for that sample
pca.data

# The X-axis tells us what percentage of the variation in the original data that PC1 accounts for
# The Y-axis tells us what percentage of the variation in the original data that PC2 accounts for
# Now the samples are labeled, so we know which ones are on the left and the right
# ★ geom_text : plot the labels, rather than dots or some other shape
# ★ theme_bw : makes the graph's background white
ggplot(data=pca.data, aes(x=X, y=Y, label=Sample)) +
  geom_text() +
  xlab(paste("PC1 - ", pca.var.per[1], "%", sep=""))+
  ylab(paste("PC2 - ", pca.var.per[2], "%", sep=""))+
  theme_bw() +
  ggtitle("My PCA Graph")

# Lastly, let's look at how to use loading scores to determine
# which genes have the largest effect on where samples are plotted in the PCA plot
# ★ The prcomp() function calls the loading scores rotation
# There are loading scores for each PC
# Here we're just going to look at the loading scores for PC1
# since it accounts for 92% of the variation in the data
loading_scores <- pca$rotation[,1]

# Genes that push samples to the left side of the graph will have large negative values
# and genes that push samples to the right will have large positive values
# Since we're interested in both sets of genes,
# we'll use the abs() function to sort based on the number's magnitude rather than from high to low
gene_scores <- abs(loading_scores)

# Now we sort the magnitude of the loading scores, from high to row
gene_score_ranked <- sort(gene_scores, decreasing=TRUE)

# Now we get the names of top 10 genes with the largest loading score magnitudes
top_10_genes <- names(gene_score_ranked[1:10])
top_10_genes

# Show the scores (and +/- sign)
# Lastly, we can see which of these genes have positive loading scores, these push the "ko" samples to the right side of the graph
# then we see which genes have negative loading scores, these push the "wt" samples to the left side of the graph
pca$rotation[top_10_genes,1]

File: statQuest/PCA.py

import pandas as pd
import numpy as np
import random as rd
import matplotlib.pyplot as plt

# There are a few Python PCA functions to choose from
# The one in sklearn is the most commonly used
from sklearn.decomposition import PCA

# Scale the data before performing PCA
from sklearn import preprocessing

# Generate an array of 100 gene names
# gene name : gene1, gene2, etc.
genes = ['gene' + str(i) for i in range(1,101)]

# Create arrays of sample names
# 5 wild type or "wt" samples
# 5 knock out or "ko" samples
wt = ['wt' + str(i) for i in range(1,6)]
ko = ['ko' + str(i) for i in range(1,6)]

# Create a dataframe
# ★ The stars "*" unack the "wt" and "ko" arrays
#    so that the column names are a single array that looks like this:
#    [wt1, wt2, wt3, wt4, wt5, ko1, ko2, ko3, ko4, ko5]
# ★ Without the stars, we would create an array of two arrays
#    and that wouldn't create 12 columns like we want:
#    [[wt1, wt2, wt3, wt4, wt5], [ko1, ko2, ko3, ko4, ko5]]
# The gene names are used for the index, which means that they are the equivalent of row names
data = pd.DataFrame(columns=[*wt,*ko], index=genes)

# Create the random data
# For each gene in the index (gene1, gene2, ...)
# create 5 values for the "wt" samples and 5 values for the "ko" samples
# The data comes from two poisson distributions
# For each gene, we select a new mean for the poisson distribution
# The means can vary between 10 and 1000
for gene in data.index:
    data.loc[gene, 'wt1':'wt5'] = np.random.poisson(lam=rd.randrange(10,1000), size=5)
    data.loc[gene, 'ko1':'ko5'] = np.random.poisson(lam=rd.randrange(10,1000), size=5)

print(data.head())
print(data.shape)

# Center and scale the data
# After centering, the average value for each gene will be 0
# After scaling, the standard deviation for the values for each gene will be 1
# ★ Notice that we are passing in the transpose of our data
#    The scale function expects the samples to be rows instead of columns
# NOTE : We use samples as columns in this example
#        because that is often how genomic data is stored
#        If you have other data, you can store it however is easiest for you
#        There's no requirement that samples be rows or columns
#        ★ Just be aware that if it is columns, you'll need to transpose it before analysis
# NOTE : The code below is just one way to use sklearn to center and scale the data
#        so that the means for each gene are 0 and the standard deviation for each gene are 1
#        ★ Alternatively, we could have used: StandardScaler().fit_transform(data.T)
#           This method is more commonly used for machine learning
#           and that's what sklearn was designed to do
# NOTE : In sklearn, variation is calculated as
#           (measurements - mean)^2 / the number of measurements
#        In R using scale() or prcomp(), variation is calculated as
#           (measurements - mean)^2 / (the number of measurements - 1)
#           -> ★ This method results in larger, but unbiased, estimates of the variation
#        The good news is that these differences do not effect the PCA analysis
#        - The loading scores and the amount of variation per principal component will be the same, either way
#        The bad news is that these differences will have a minor effect the final graph
#        - This because the coordinates on the final graph come from multiplying the loading scores by the scaled values
scaled_data = preprocessing.scale(data.T)

# Create a PCA object
# Rather than just have a function that does PCA and returns results
# sklearn uses objects that can be trained using one dataset and applied to another dataset
pca = PCA()

# Since we're only using PCA to explore one dataset
# (and not using PCA in a machine learning setting),
# the additional steps are a little tedious,
# but they set us up for the machine learning topics

# This is where we do all of the PCA math
# i.e. Calculate loading scores and the variation each principal component accounts for
pca.fit(scaled_data)

# Generate coordinates for a PCA graph based on the loading scores and the scaled data
pca_data = pca.transform(scaled_data)

# Draw a graph
# We'll start with a scree plot to see how many principal components should go into the final plot

# First, calculate the percentage of variation that each principal component accounts for
per_var = np.round(pca.explained_variance_ratio_* 100, decimals=1)

# Create labels for the scree plot
# These are "PC1", "PC2", etc
labels = ['PC' + str(x) for x in range(1, len(per_var)+1)]

# Create bar plot
# Almost all of the variation is along the first PC,
# so a 2-D graph, using PC1 and PC2, should do a good job representing the original data
plt.bar(x=range(1, len(per_var)+1), height=per_var, tick_label=labels)
plt.ylabel('Percentage of Explained Variance')
plt.xlabel('Principal Component')
plt.title('Scree Plot')
plt.show()

# To draw a PCA plot, we'll first put the new coordinates,
# created by pca.transform(scaled.data), into a nice matrix
# whre the rows have sample labels and the columns have PC labels
pca_df = pd.DataFrame(pca_data, index=[*wt, *ko], columns=labels)

plt.scatter(pca_df.PC1, pca_df.PC2)
plt.title('My PCA Graph')
plt.xlabel('PC1 - {0}%'.format(per_var[0]))
plt.ylabel('PC2 - {0}%'.format(per_var[1]))

# This loop adds sample names to the graph
for sample in pca_df.index:
    plt.annotate(sample, (pca_df.PC1.loc[sample], pca_df.PC2.loc[sample]))

# Display the graph
plt.show()

# Result
# The "wt" samples clustered on the left side, suggesting that they are correlated with each other
# The "ko" samples clustered on the right side, suggesting that they are correlated with each other
# And the separation of the two clusters along the x-axis suggests that
# "wt" samples are very different from "ko" samples

# Lastly, let's look at the loading scores for PC1 to determine which genes had the largest influence
# on separating the two clusters along the x-axis
# We'll start by creating a pandas "Series" object with the loading scores in PC1
# NOTE : The PCs are zero-indexed, so PC1 = 0
loading_scores = pd.Series(pca.components_[0], index=genes)

# Sort the loading scores based on their magnitude (absolute value)
sorted_loading_scores = loading_scores.abs().sort_values(ascending=False)

# Get the names of the top 10 indexes
top_10_genes = sorted_loading_scores[0:10].index.values

# Print out the top 10 gene names and their corresponding loading scores
# These values are super similar, so a lot of genes played a role in separating the samples,
# rather than just one or two
print(loading_scores[top_10_genes])
